File: empresa.c
/*
 * File:  empresa.c
 * Autor:  Laercio Costa Rodrigues
 * Descricao: Teste Técnico Intelitrader
*/

#include "empresa.h"


int representantes, website, app_android, app_iphone;


/* Registra produtos no estoque */
void registra_produto(lista_estoque* le, int codigo, int qt_inicial, int qt_co) {
    pt_no_estoque novo_no = cria_no_estoque(codigo, qt_inicial, qt_co);
    adiciona_estoque_fim(le, novo_no);
}


/* Registra produtos no venda */
void registra_venda(lista_venda* lv, int codigo, int qt_vendida,
                             int situacao_venda, int canal_venda){
                                 
    pt_no_venda novo_no = cria_no_venda(codigo, qt_vendida, situacao_venda, canal_venda);
    adiciona_venda_fim(lv, novo_no);
}


/* Inicializa valores de vendas por canal */
void inicializa_vendas_por_canal() {
    representantes = 0;
    website = 0;
    app_android = 0;
    app_iphone = 0;
}


/* Incrementa Venda pelo canal Representantes */
void incrementa_representantes(int qtVendida) {
    representantes += qtVendida;
}


/* Incrementa Venda pelo canal WebSite */
void incrementa_website(int qtVendida) {
    website += qtVendida;
}


/* Incrementa Venda pelo canal App Android */
void incrementa_app_android(int qtVendida) {
    app_android += qtVendida;
}


/* Incrementa Venda pelo canal App Iphone */
void incrementa_app_iphone(int qtVendida) {
    app_iphone += qtVendida;
}


/* Retorna Venda pelo canal Representantes */
int retorna_representantes(){
    return representantes;
}


/* Retorna Venda pelo canal WebSite */
int retorna_website(){
    return website;
}


/* Retorna Venda pelo canal App Android */
int retorna_app_android(){
    return app_android;
}


/* Retorna Venda pelo canal App Iphone */
int retorna_app_iphone(){
    return app_iphone;
}

File: Empresa.java
import java.util.*;

public class Empresa {

   private Map<Integer, Estoque> estoqueMapa;
   private Map<Integer, Venda> vendaMapa;
   private List<String> divergenciasLista;
   private List<String> transfereLista;
   private int representantes = 0, webSite = 0;
   private int appAndroid = 0, appIphone = 0;

   public Empresa() {
      estoqueMapa = new LinkedHashMap<Integer, Estoque>();
      vendaMapa = new LinkedHashMap<Integer, Venda>();
      divergenciasLista = new ArrayList<String>();
      transfereLista = new ArrayList<String>();
   }

   /* Incrementa Venda pelo canal Representantes */
   public void incrementaRepresentantes(int representantes) {
      this.representantes += representantes;
   }

   /* Incrementa Venda pelo canal WebSite */
   public void incrementaWebSite(int webSite) {
      this.webSite += webSite;
   }

   /* Incrementa Venda pelo canal App Android */
   public void incrementaAppAndroid(int appAndroid) {
      this.appAndroid += appAndroid;
   }

   /* Incrementa Venda pelo canal App Iphone */
   public void incrementaAppIphone(int appIphone) {
      this.appIphone += appIphone;
   }

   public int getRepresentantes() { return representantes; }

   public int getWebSite() { return webSite; }

   public int getAppAndroid() { return appAndroid; }

   public int getAppIphone() { return appIphone; }


   void registraProduto(int codigo, int qtInicial, int qtCO) {
      estoqueMapa.put(codigo, new Estoque(codigo, qtInicial, qtCO));
   }

   void registraVenda(int codigo, int qtVendida , int situacaoVenda, int canalVenda ) {
      vendaMapa.put(codigo, new Venda(codigo, qtVendida, situacaoVenda, canalVenda));
   }

   void registraDivergencia(String linha){
      divergenciasLista.add(linha);
   }

   public boolean contemEstoque(int codigo) {
      return estoqueMapa.containsKey(codigo);
   }

   public boolean contemVenda(int codigo) {
      return vendaMapa.containsKey(codigo);
   }

   void incrementaVenda(int codigo, int qtVendida) {
      Venda v = vendaMapa.get(codigo);
      if (v != null)
         v.incrementaQtVendida(qtVendida);
   }

   public Collection<Estoque> getEstoqueMapa() {
      return Collections.unmodifiableCollection(estoqueMapa.values());
   }

   public Collection<String> getDivergenciaLista() {
      return Collections.unmodifiableCollection(divergenciasLista);
   }

   public Collection<String> getTransfereLista() {
      return Collections.unmodifiableCollection(transfereLista);
   }

   void criaTransfere() {

      for (Estoque e : getEstoqueMapa()) {
         Venda v = vendaMapa.get(e.getCodigo());
         String linha = "";

         if (v != null) {
            int necessario = 0, transferir;

            if (e.getQtInicial() - v.getQtVendida() - e.getQtCO() < 0)
               necessario = e.getQtCO() - (e.getQtInicial() - v.getQtVendida());
               
            if (necessario > 1 && necessario < 10)
               transferir = 10;
            else
               transferir = necessario;

            linha += e.getCodigo(); // Coluna 1 -> Produto
            linha += String.format("%8d", e.getQtInicial()); // Coluna 2 -> QtCO
            linha += String.format("%7d", e.getQtCO()); // Coluna 3 -> QtMin
            linha += String.format("%10d", v.getQtVendida()); // Coluna 4 -> QtVendas
            linha += String.format("%11d", e.getQtInicial() - v.getQtVendida()); // Coluna 5 -> Estq.após Vendas
            linha += String.format("%9d", necessario); // Coluna 6 -> Necess.
            linha += String.format("%12d", transferir); // Coluna 7 - > Transf. de Arm p/ CO

            transfereLista.add(linha);
         }
      }
   }

   // Importa dados do Arquivo produtos.txt
   void importaEstoque(String txtfile) {
         ParserImportacao parser = new ParserImportacao(this);
         parser.importaEstoque(txtfile);
      }

   // Importa dados do Arquivo vendas.txt
   void importaVenda(String txtfile) {
      ParserImportacao parser = new ParserImportacao(this);
      parser.importaVenda(txtfile);
   }

   // Exporta dados para o Arquivo transfere.txt
   void exportaTransfere(String txtfile) {
      criaTransfere();
      ParserExportacao parser = new ParserExportacao(this);
      parser.exportaTransfere(txtfile);
   }

   // Exporta arquivo DIVERGENCIAS.TXT
   void exportaRelatorioDivergencias(String txtfile) {
      ParserExportacao parser = new ParserExportacao(this);
      parser.exportaRelatorioDivergencias(txtfile);
   }

   // Exporta dados para o Arquivo TOTCANAIS.txt
   void exportaRelatorioPorCanais(String txtfile) {
      ParserExportacao parser = new ParserExportacao(this);
      parser.exportaRelatorioPorCanais(txtfile);
   }
}